src/games/erdos_renyi.c

```diff
@@ -244,6 +244,120 @@ igraph_error_t igraph_erdos_renyi_game_gnm(
     return IGRAPH_SUCCESS;
 }
 
+igraph_error_t igraph_erdos_renyi_game_gnm_multi(
+    igraph_t *graph, igraph_integer_t n, igraph_integer_t m,
+    igraph_bool_t directed, igraph_bool_t loops, igraph_bool_t multi
+) {
+
+    igraph_integer_t no_of_nodes = n;
+    igraph_integer_t no_of_edges = m;
+    igraph_real_t no_of_nodes_real = (igraph_real_t) no_of_nodes;   /* for divisions below */
+    igraph_vector_int_t edges = IGRAPH_VECTOR_NULL;
+
+    if (n < 0) {
+        IGRAPH_ERROR("Invalid number of vertices.", IGRAPH_EINVAL);
+    }
+    if (m < 0 || m > IGRAPH_ECOUNT_MAX) {
+        IGRAPH_ERROR("Invalid number of edges.", IGRAPH_EINVAL);
+    }
+
+    if (m == 0.0 || no_of_nodes == 0) {
+        IGRAPH_CHECK(igraph_empty(graph, n, directed));
+        igraph_vector_int_destroy(&edges);
+        IGRAPH_FINALLY_CLEAN(1);
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 0);
+    IGRAPH_CHECK(igraph_vector_int_reserve(&edges, no_of_edges * 2));
+
+    igraph_integer_t i;
+    igraph_real_t maxedges;
+    if(multi) {
+        maxedges = (igraph_real_t) IGRAPH_ECOUNT_MAX;
+    } else {
+        maxedges = n;
+        if(directed && loops) {
+            maxedges *= n;
+        } else if(directed && !loops) {
+            maxedges *= n-1;
+        } else if(!directed && loops) {
+            maxedges *= (n+1) / 2.0;
+        } else {
+            maxedges *= (n-1) / 2.0;
+        }
+    }
+    if(no_of_edges > maxedges) {
+        IGRAPH_ERROR("Too many edges requested compared to the number of vertices.", IGRAPH_EINVAL);
+    }
+    
+    if(multi) {
+
+        RNG_BEGIN();
+        for(i = 0; i < no_of_edges; i++) {
+            igraph_integer_t from, to;
+            from = RNG_INTEGER(0, no_of_nodes - 1);
+            to = RNG_INTEGER(0, no_of_nodes - 1);
+            if(!loops && from == to) {
+                to = (to + 1) % (no_of_nodes - 1);
+            }
```

Comment:
The handling of loops, as done here, is incorrect, and does not sample graphs uniformly.

Have an entirely separate conditional branch for the case with loops, and use the following algorithm:

 - Sample `from` from $[0, n-1]$, where $n$ is the vertex count.
 - Sample `to` from $[0,n-2]$, and if `to == from` then set `to = n-1`.

Basically, you need to sample `to` uniformly from not $n$, but $n-1$ distinct values. When you do the modulo operation, as implemented now, there is a chance that it wraps around to `0`, placing more weight on `0` than on other values.